### ISSUE TITLE
Support TypeScript `import = require` syntax

### ISSUE THREAD

**Issue by j-f1** [benmosher/eslint-plugin-import#1244]
Currently this project only parses regular import statements as import statements. However, TypeScript also supports another syntax:

```ts
// TSImportEqualsDeclaration
import foo = require('bar')
//           ^^^^^^^^^^^^^^
// TSExternalModuleReference
```

Would it be possible for the plugin to check these imports as well?

**Comment by ljharb** [comment]
TypeScript's module system is broken unless synthetic imports and esModuleInterop are enabled. With these settings, I don't believe there's any use case for the `import = require` syntax. Can you confirm?

**Comment by ljharb** [comment]
For example, a common thing people do without those options is `import * as React from ‘react’`, which is incorrect - React only has a default export, and import * brings in a frozen Module Namespace object, which TS does not by default.

**Comment by j-f1** [comment]
Ah, interesting.

Still, since people still use TypeScript without `esModuleInterop` and the syntax is valid, there doesn’t seem to be any harm in adding support for it.

**Comment by bradzacher** [comment]
Ultimately it shows a weakness of using CJS to represent es6 modules.

```JS
const exp = { a, b, c };
module.exports = exp;

// does this mean
export default exp;
// or
export { a, b, c };
```

The intent is definitely the former, but based on how the transpilation works, reality is the latter.

**Comment by manuth** [comment]
I'm also waiting for this plugin to suport `import x = require("x");` expressions.
While it might be quite confusing what's exactly the difference between `import * as x from "x";` (which is supported by this module) and `import x = require("x");` (which is not supported by this plugin) I try to explain it very briefly.

The module `assert`, for example, directly exports a method which is made for asserting truthy values:
```js
const assert = require("assert");
assert(true); // does not throw
assert(false); //does throw
```

And here comes the clue:
I've found a quite good explanation on [StackOverflow](https://stackoverflow.com/a/35706271) which might help you understanding.
While this here is perfectly valid typescript-code:
***index.ts***
```ts
import assert = require("./MyAssert");
assert(true);
```

This here is not:
```ts
import * as assert from "./MyAssert";
import * as MyClass from "./MyClass";

assert(true);
new MyClass();
```

Because while `import x = require("x");` assigns the exported object to `x`, `import * as x from "x"` fetches all exported members and assigns them to `x`.

So there certainly is a reason why `import [...] = require[...]` exists and it'd be very awesome if this plugin would support said sort of import.

Thanks for this great plugin ^^

**Comment by ljharb** [comment]
@manuth that's not quite correct; if you use esModuleInterop and synthetic imports (as `tsc --init` enables), then you never need `import * as` in TS (which is only useful in regular JS for laziness or metaprogramming), nor `import =` (which is nonstandard syntax). The solution is to fix your TS config so TS's module system isn't broken.

**Comment by manuth** [comment]
It is valid typescript code and therefore it would be quite nice if this plugin could handle it. That's all people commenting here were asking for.

**Comment by ljharb** [comment]
I totally agree, which is why i put the "help wanted" label on this issue. I'm saying that regardless, the best solution is to avoid using this syntax in the first place, and I wanted to correct your explanation about it.

**Comment by manuth** [comment]
Thanks for your help then 😄 
I'll try to get some PR done to get TypeScript's "import equals"-references to work.

**Comment by ExE-Boss** [comment]
Note that `import/first` now has conflicts with `import/order` when **TypeScript**’s `import =` is used.

**Comment by ljharb** [comment]
@ExE-Boss thanks for the report! a PR with a fix, or a PR with a failing test case, or an issue would be great (in that order ;-) ) :-D
